docs/advanced/kotlin-codegen.md

```diff
@@ -0,0 +1,85 @@
+
+The generated classes in this mode wrap each field in a `Supplier<T>`, or in Kotlin, `() -> T`. This is very similar to lazy values, but with some differences. What this allows us to do is defer the evaluation of a field until it is accessed, either on the server for serialization, or on the client for consumption. Luckily, the Kotlin closure syntax allows us a compact way to express these, simply by wrapping them in curly braces:
+
+```kotlin
+val series = Series(
+    title = { "Stranger Things" },
+    actors = { listOf("Millie Bobby Brown", "Finn Wolfhard", "Winona Ryder", "David Harbour") },
+)
+```
+
+Alternatively, we also generate builder methods where you can construct these objects as such:
+
+```kotlin
+val series = Series.Builder()
+    .withTitle("Stranger Things")
+    .withReleaseDate(2016)
+    .withEndDate(2024)
+```
+
+Note that in each of these examples, we're only populating partial objects, which is frequently the case in GQL. If the user were to access either the release dates in the first example, or the list of actors in the second, an exception would be thrown indicating that those fields are not populated.
+
+We can also specify explicit null response values for when we want to return an explicit null value for a field that was requested:
+
+
+```kotlin
+val series = Series(
+    title = { "Black Mirror" },
+    releaseDate = { 2011 },
+    endDate = { null },
+)
+```
+
+In this example, `endDate` would be a nullable Kotlin field, whereas `title` would be non-nullable. Also note that the supplier closure is not exposed when accessing the fields. These are all properties that fetch the value when accessed (or throw an exception if it was not populated)
+
+```kotlin
+val title: String = series.title
+val releaseDate: Int = series.releaseDate
+val endDate: Int? = series.endDate
+```
+
+## Query Projections
+
+Kotlin supports [function literals with a receiver](https://kotlinlang.org/docs/lambdas.html#function-literals-with-receiver) and this allows us to mimic the GQL query syntax, directly in the language. An advantage of writing queries directly in the language is that when the schema changes, any incompatibilities will show up as compile time errors, and intellisense can guide users to craft queries. Additionally, nested projections are nested in the query, just as they appear in a GQL query. For example:
```

Comment:
Don't use "intellisense"; make it generic like "your IDE"

Reply:
huh - I had no idea that was a msft thing. Guess I must have picked it up there & have been using it for any IDE 🤔

Reply:
Fixed